tests/api/test_payment.py

```diff
@@ -454,19 +460,120 @@ def test_query_payments(payment_dummy, permission_manage_orders, staff_api_clien
-def test_query_payment_client_token(mock_get_client_token, user_api_client):
+@pytest.fixture
+def braintree_customer_id():
+    return "1234"
+
+
+def test_store_payment_gateway_meta(customer_user, braintree_customer_id):
+    gateway_name = PaymentGatewayEnum.BRAINTREE.name
+    META = {"gateways": {gateway_name.upper(): {"customer_id": braintree_customer_id}}}
+    store_id_for_payment_gateway(customer_user, gateway_name, braintree_customer_id)
+    assert customer_user.private_meta == META
+    customer_user.refresh_from_db()
+    assert (
+        extract_id_for_payment_gateway(customer_user, gateway_name)
+        == braintree_customer_id
+    )
+
+
+def test_query_payment_client_token(mocker, user_api_client):
     query = """
     query paymentClientToken($gateway: GatewaysEnum) {
         paymentClientToken(gateway: $gateway)
     }
     """
     example_token = "example-token"
-    mock_get_client_token.return_value = example_token
+    mock_get_client_token = mocker.patch(
+        "saleor.graphql.payment.resolvers.gateway_get_client_token",
+        return_value=example_token,
+    )
     variables = {"gateway": PaymentGatewayEnum.BRAINTREE.name}
     response = user_api_client.post_graphql(query, variables)
 
     content = get_graphql_content(response)
-    assert mock_get_client_token.called_once_with(PaymentGatewayEnum.BRAINTREE.name)
+    mock_get_client_token.assert_called_once_with("braintree")
     token = content["data"]["paymentClientToken"]
     assert token == example_token
+
+
+@pytest.fixture
+def token_config_with_customer(braintree_customer_id):
+    return TokenConfig(customer_id=braintree_customer_id)
+
+
+@pytest.fixture
+def set_braintree_customer_id(customer_user, braintree_customer_id):
+    gateway_name = "braintree"
+    store_id_for_payment_gateway(customer_user, gateway_name, braintree_customer_id)
+    return customer_user
+
+
+def test_use_customer_id_from_meta_on_client_token_generation(
+    mocker,
+    token_config_with_customer,
+    braintree_customer_id,
+    set_braintree_customer_id,
+    user_api_client,
+):
+    query = """
+    query paymentClientToken($gateway: GatewaysEnum) {
+        paymentClientToken(gateway: $gateway)
+    }
+    """
+    TOKEN = "sample_token"
+    mock_get_token = mocker.patch(
+        "saleor.graphql.payment.resolvers.gateway_get_client_token",
+        return_value=TOKEN,
+        autospec=True,
+    )
+    variables = {"gateway": PaymentGatewayEnum.BRAINTREE.name}
+    response = user_api_client.post_graphql(query, variables)
+
+    mock_get_token.assert_called_once_with(
+        "braintree", TokenConfig(customer_id=braintree_customer_id)
+    )
+    content = get_graphql_content(response)
+    assert content["data"]["paymentClientToken"] == TOKEN
+
+
+@pytest.fixture
+def enable_braintree(settings):
+    settings.CHECKOUT_PAYMENT_GATEWAYS = {"braintree": None}
+    return settings
+
+
+def test_list_payment_sources(
```

Comment:
You might want to check that it doesn't work for anonymous users and that a user cannot access other users' data.

Reply:
Separate test for resolver decorator seems excessive, query fetches data for logged user so checking for other users data is pointless and even is possible does not provide enough information to do anything useful with it.